### PR TITLE
python(folds): fold whole functions not only body

### PR DESCRIPTION
**Pull request by theHamsta**
Fold queries of Python were an issue in this query https://github.com/nvim-treesitter/nvim-treesitter/issues/1441

I think we should streamline the Python fold queries with the other languages. I think this will be also beneficial for everyone liking the current behavior as then all languages could be folded with first-line visible. Right, know "omitting the first line from folding" would require a different logic for Python and for C++.

### CODE REVIEW

**Comment by theHamsta** [comment]
We can of course also merge this after the PR for the different folding modes but at some time we have to enable both options.

**Comment by stsewd** [comment]
I think we had this conversation before, we have people on both sides. I think till neovim implements column based folding there isn't a good answer, but also if that's implemented we would need to change back to capture the body. I think a better solution is to have `fold.inner` and `fold.outer` captures, and a setting to choose between one or both.

**Comment by theHamsta** [comment]
> , but also if that's implemented we would need to change back to capture the body. I

Why? Is it because you ultimate want to have folds in the style of `int main() {...}` with `...` being the column folded region`? But this doesn't apply to Python where bodies reside entirely in their own line.

> I think we had this conversation before, we have people on both sides

But your argument always was that you don't like how Neovim was displaying the fold and that you want to see the first line. This has nothing to do with the queries and can be described for all languages in Lua. There queries only give the information `where is the function?` not how this should be handled.

Can you describe why
- why your previous PR wouldn't resolve the situation? The only problem case I would expect when there would be multi-line function headers. Wouldn't still be good that 
```python
def function(a,
                    b,
                    c):
      pass
```
would be folded as
```python
def function(a,
<fold text here>
```
It wouldn't occur that often and would still represent the function in  a compact manner. What is the problem with this implementation that could be implemented for all languages without any changes of the queries.
- why a one-line folding *with syntax highlighting* wouldn't resolve the solution? It would look like your Atom screen shot https://github.com/nvim-treesitter/nvim-treesitter/pull/390#issuecomment-721454440

**Comment by theHamsta** [comment]
I think `@fold.outer`/`@fold.inner` could be the solution when we still allow normal `@fold` in order to not have to refactor all folds.

There could be two modes:

- mode 1: fold at `@fold.outer` and set foldtext to `<beginning of fold outer text> { ... }` where `...` is fold inner omitted
- mode 2: fold at `@fold.inner` 

when there is only `@fold` it would just use current behavior.

**Comment by theHamsta** [comment]
Will change this to the outer/inner proposal

**Comment by JoZie** [comment]
>Why? Is it because you ultimate want to have folds in the style of int main() {...} with ... being the column folded region`? But this doesn't apply to Python where bodies reside entirely in their own line.

If you set the `foldtext` correctly this folding should be currently possible.
The real challenge is to get the highlighting as expected.
AFAIK a folded line is assigned to the  `Fold` highlight group as a whole. I'm unsure how to change that.

**Comment by theHamsta** [comment]
My idea was to have a setting whether the users want to fold at the definition or the definition body. Depending on this setting. Folding would either pick `@fold` and `@fold.outer` captures or `@fold` and `@fold.inner`.

**Comment by wookayin** [comment]
I like the proposal on configurability between `@fold.inner` and `@fold.outer`. However, I believe this PR could still be merged in its current form (with some additional tweaks if necessary), because *all other* languages are folding the whole definition region. Python should not be an exception (for default settings). Can @theHamsta and maintainers please revisit this discussion again? Later we could make another thread for improving and adding the customization feature.

**Comment by theHamsta** [comment]
I wouldn't use fold.inner/fold.out. I assume that also the desired behavior for the current behavior is just to leave out the first line of the function to make it visible. We could implement it as an option here in this plugin that just optionally offsets 1 line or create a window option for Neovim.

**Comment by theHamsta** [comment]
Let me add the functionality for @stsewd style folding this weekend (like in @stsewd original PR). I still think that it would be better to have folding formatting options in core, so that you can also have other styles of folding with the same defintions and that you could also fold at the function signature with only the body being hidden. But it's definitely easier to implement the first workaround here

**Comment by wookayin** [comment]
That could be a nice improvement for not only python but also other languages. Sounds like the foldexpr function will have some options and logics implemented in lua. But for now we can have this, being consistent with other languages, and IMHO should be the default behavior unless we go with the second style for all other languages as well.

**Comment by WieeRd** [comment]
Is there a reason this PR is not getting merged or just not noticed by the team?  
Always wondered why only Python had inconsistent folding method and just discovered this PR.

**Comment by theHamsta** [comment]
@clason @stsewd @kyazdani42 I would also love to merge this. I'm creating right now a separate PR that would allow to opt-in for the old behavior for all languages.

**Review by kyazdani42 — APPROVED** [review]
go for it then

**Pull request merged**
